### ISSUE TITLE
New feature: Project specific rules

### ISSUE THREAD

**Issue by joxeankoret** [joxeankoret/diaphora#110]
I'm opening this issue to publicly discuss how to add to Diaphora the ability to have "hooks" for both the export and diffing processes so reversers can define which rows are exported, how, apply modifications... as well as determine what is a good match, apply their own rules specific to each project...

The idea is as easy as the following: in the export dialog, or via a command line argument or environment variable, the reverser specifies a Python script that will be used to have "hooks" for various Diaphora's processes. For example: exporting rows. Let's say that you only want to export the syscall and ioctl handlers from a kernel. There is no way in the current version of Diaphora that let's you export exclusively such functions. However, one can write a project specific script and do something like the following:

```
class CMyProjectHooks:
  def before_export_function(self, ea):
    name = GetFunctionName(ea)
    return name.startswith("syscall_") or name.startswith("ioctl_")

```

The method ```CMyProjectHooks.before_export_function()``` would be called for each function to export and there the reverser can specify which functions she or he wants.

Another example that shows the usefulness of such hooks. Let's say that the reverser is diffing a project against a target that is very old and many things changed. Naturally, the results will be very bad. However, the reverser finds out that there is a specific **sub-string** that, when found, matches a function 1 to 1. The reverser could write a code like the following one:

```
class CMyProjectHooks:
  def __init__(self, diaphora):
    self.diaphora = diaphora

  def find_best_matches(self, db):
   """ Called for finding best matches. """
    sql = "select column_names from both databases where substr(main.name) = substr(diff.name)"
    self.diaphora.add_best_matches_from_query(sql)

  def find_partial_matches(self, db):
   """ Called for finding partial matches. """
    sql = "select column_names from both databases where substr(main.name) = substr(diff.name)"
    self.diaphora.add_partial_matches_from_query(sql)

```

And these are my ideas for now. This is subject to discussion as I'm currently in the design phase and I would like to know what people think about it and what ideas you have.

**Comment by AGG2017** [comment]
Great features! Can't wait to test it, especially the diffing part. I constantly have to add my functions after each new version of diaphora. This way the only missing part for me will be to add sub-menu items for the result windows with some extra functions when trying to understand the difference by visual comparing (mostly I would like to have clean assembly diff).

**Comment by joxeankoret** [comment]
I'm thinking about possible ways to extend Diaphora's functionality. Maybe using some short of plugins. If you have some idea, I'm all ears!

**Comment by AGG2017** [comment]
Python is totally not my force so I cannot propose any good enough solution. I just thought I can extend your class by adding more functions and at key points your original class can check for reserved function names and if they are defined you can call them to extend the current functionality. Maybe it is the same as what you plan to do but you can check also for key names of extension lists for example of additional action handler names/commands to add to the menu if they are defined in the extended class. Not sure if this can be done easily in Python.

**Comment by joxeankoret** [comment]
It can be done easily. Let me think over it.

**Comment by joxeankoret** [comment]
And finally fixed with d43bb285a0af6a0cb2e8b6516ad0cffe886c8595
